sqld/src/query_analysis.rs

```diff
@@ -62,77 +62,70 @@ impl StmtKind {
-        | "cache_spill"
-        | "cell_size_check"
-        | "checkpoint_fullfsync"
-        | "defer_foreign_keys"
-        | "encoding"
-        | "fullfsync"
-        | "hard_heap_limit"
-        | "journal_mode"
-        | "journal_size_limit"
-        | "legacy_alter_table"
-        | "locking_mode"
-        | "max_page_count"
-        | "mmap_size"
-        | "page_size"
-        | "query_only"
-        | "read_uncommitted"
-        | "recursive_triggers"
-        | "reverse_unordered_selects"
-        | "schema_version"
-        | "secure_delete"
-        | "soft_heap_limit"
-        | "synchronous"
-        | "temp_store"
-        | "threads"
-        | "trusted_schema"
-        | "user_version"
-        | "wal_autocheckpoint"
 
-            if body.is_none() =>
-        {
-            true
-        }
-        // changes the state of the connection, and can't be allowed rn:
-        "case_sensitive_like" | "ignore_check_constraints" | "incremental_vacuum" 
-            // TODO: check if optimize can be safely performed
-            | "optimize"
-            | "parser_trace"
-            | "shrink_memory"
-            | "wal_checkpoint"
-
-            => {
-            false
+    fn pragma_kind(name: &QualifiedName, body: Option<&PragmaBody>) -> Option<Self> {
+        let name = name.name.0.as_str();
+        match name {
+            // always ok
+            "foreign_key" | "foreign_key_list" | "foreign_key_check" | "collation_list"
+            | "compile_options" | "data_version" | "database_list" | "freelist_count"
+            | "function_list" | "index_list" | "index_xinfo" | "integrity_check"
+            | "legacy_file_format" | "page_count" | "pragma_list" | "quick_check" | "stats" | "table_info" | "table_list" | "table_xinfo" => Some(Self::Read),
+            // ok without args
+            "analysis_limit"
+            | "application_id"
+            | "auto_vacuum"
+            | "automatic_index"
+            | "busy_timeout"
+            | "cache_size"
+            | "cache_spill"
+            | "cell_size_check"
+            | "checkpoint_fullfsync"
+            | "defer_foreign_keys"
+            | "encoding"
+            | "fullfsync"
+            | "hard_heap_limit"
+            | "journal_mode"
+            | "journal_size_limit"
+            | "legacy_alter_table"
+            | "locking_mode"
+            | "max_page_count"
+            | "mmap_size"
+            | "page_size"
+            | "query_only"
+            | "read_uncommitted"
+            | "recursive_triggers"
+            | "reverse_unordered_selects"
+            | "schema_version"
+            | "secure_delete"
+            | "soft_heap_limit"
+            | "synchronous"
+            | "temp_store"
+            | "threads"
+            | "trusted_schema"
+            | "user_version"
+            | "wal_autocheckpoint" => {
+                Some(match body {
+                    Some(_) => Self::Write,
+                    None => Self::Read,
```

Comment:
this is not correct. All of those are read-only if they don't have params. If they have params, then they write, and we can't support that currently, and without careful consideration

Reply:
*`encoding` would be special-cased too, since it's write only when the db hasn't been created yet, which is never the case in sqld

Reply:
but notice that: if you set the `foreign_key`, then this is a write and is executed on the primary, now, if you read `foreign_key`, it's categorized as read, and read from the replica, which has a different state.

So all pragmas need to be executed on the primary. Let's add a PRAGMA category, and send that to the primary

Reply:
> So all pragmas need to be executed on the primary.

I think this is too strong a statement. Some pragmas behave like that - probably all the ones that can either set or read parameters. But there's also a class of pragmas that are only readonly and it's fine to send them to replicas - e.g. ones that return schema information. I believe these should be allowed to be executed on replicas too

Reply:
A special category for pragmas that are reads-but-should-happen-on-primary sounds good in general though. I think it's not really different from reading and writing without a transaction, just a snapshot isolation thing, but no harm in making pragmas resilient to that

Reply:
@MarinPostma ultimately the current way is safer, so let's go with that *until, if ever* somebody with a read-only token complains. Meanwhile I'll just rework this PR to have the same semantics as yours, but with the new structure, because it will make it easier for me to try and figure out a correct solution for RO pragmas as a low priority weekend project, wdyt?

Reply:
oor maybe I'll just not resist myself and add a short list of pragmas I'm **sure** are reads, and let you review.